### ISSUE TITLE
False positive on return type

### ISSUE THREAD

**Issue by alexei** [python/mypy#17445]
I found an instance where mypy is not behaving the way a developer would expect:

```python
import random


def f1() -> int:
    if bool(random.getrandbits(1)):
        retval = 1
    return retval


def f2() -> int:
    if False:
        retval = 1
    return retval
```

mypy reports on line 13: "error: Cannot determine type of "retval"  [has-type]", but not on line 7 as well. I would expect it to treat `retval` as an undefined variable as the condition can also be false - there's no way to tell.

Gist URL: https://gist.github.com/mypy-play/33d1f016c63e3c72dcaf1e89256b4af0
Playground URL: https://mypy-play.net/?mypy=latest&python=3.12&gist=33d1f016c63e3c72dcaf1e89256b4af0

**Comment by alexei** [comment]
Note I originally found this on a function `f1` like this:

```python
class A: ...


class B: ...


def c1(): ...


def c2(): ...


def c3(): ...


def r1():
    if c1():
        return A()


def r2():
    return B()


def f1() -> A | B:
    if c3():
        r = r1()
    if not r or not r.foo:
        r = r2()
    return r


def f2() -> A | B:
    r = None
    if c3():
        r = r1()
    if not r or not r.foo:
        r = r2()
    return r
```

`f1` passed the static checks, but in production it failed because all conditions were false. Then I fixed it to `f2` which, to my rurprise, mypy still approves of.

Gist URL: https://gist.github.com/mypy-play/90552823a3fd2bc93ab632c652e55720
Playground URL: https://mypy-play.net/?mypy=latest&python=3.12&gist=90552823a3fd2bc93ab632c652e55720

**Comment by alexei** [comment]
> Note I originally found this on a function `f1` like this:

I now realise why this one passed: it's in the condition `if not r or ...`

**Comment by hauntsaninja** [comment]
You can enable the possibly-undefined error code https://mypy.readthedocs.io/en/stable/error_code_list2.html#warn-about-variables-that-are-defined-only-in-some-execution-paths-possibly-undefined

**Comment by alexei** [comment]
Thank you, @hauntsaninja
